File: src/commonMain/kotlin/com/camackenzie/exvi/core/model/Muscle.java
/*
 * To change this license header, choose License Headers in Project Properties.
 * To change this template file, choose Tools | Templates
 * and open the template in the editor.
 */
package com.camackenzie.exvi.core.model;

import com.camackenzie.exvi.core.model.EnumUtils;
import java.util.HashSet;

/**
 *
 * @author callum
 */
public enum Muscle {
    PECTORALIS_MAJOR("pecs"),
    TRAPEZIUS("traps"),
    RHOMBOIDS(),
    LATISSIMUS_DORSI("lats", "latissimus"),
    ERECTOR_SPINAE("erector"),
    BICEPS_BRACHII(),
    BICEPS_BRACHIALIS(),
    TRICEPS("tricep"),
    FOREARMS("forearm"),
    ANTERIOR_HEAD(),
    LATERAL_HEAD(),
    POSTERIOR_HEAD(),
    QUADRICEPS("quads", "quad"),
    ABDUCTORS(),
    ADDUCTORS(),
    GLUTES(),
    IT_BAND(),
    HIP_FLEXORS(),
    GASTROCNEMIUS(),
    SOLEUS(),
    RECTUS_ABDOMINIS(),
    TRANSEVERSE_ABDOMINIS(),
    INTERNAL_OBLIQUES(),
    EXTERNAL_OBLIQUES(),
    HAMSTRINGS(new Muscle[]{ABDUCTORS, ADDUCTORS, IT_BAND}),
    BICEPS(new Muscle[]{BICEPS_BRACHIALIS, BICEPS_BRACHII}),
    LOWER_BACK(ERECTOR_SPINAE),
    UPPER_BACK(new Muscle[]{TRAPEZIUS, RHOMBOIDS}),
    NECK(),
    OBLIQUES(new Muscle[]{INTERNAL_OBLIQUES, EXTERNAL_OBLIQUES}),
    PALMAR_FASCIA("grip"),
    PLANTAL_FASCIA("feet"),
    BACK(new Muscle[]{LOWER_BACK, UPPER_BACK, LATISSIMUS_DORSI}),
    CHEST(new Muscle[]{PECTORALIS_MAJOR}),
    ARMS(new Muscle[]{BICEPS, PALMAR_FASCIA, TRICEPS, FOREARMS}),
    SHOULDERS(new Muscle[]{ANTERIOR_HEAD, LATERAL_HEAD, POSTERIOR_HEAD}),
    LEGS(new Muscle[]{QUADRICEPS, GLUTES, HAMSTRINGS}),
    CALVES(new Muscle[]{GASTROCNEMIUS, SOLEUS}),
    ABS(new Muscle[]{RECTUS_ABDOMINIS, OBLIQUES, TRANSEVERSE_ABDOMINIS});

    private final String[] altNames;
    private final HashSet<Muscle> subMuscles = new HashSet<>();
    private final HashSet<Muscle> superMuscles = new HashSet<>();

    private Muscle(Muscle[] subMuscles, String... altNames) {
        this.altNames = altNames;
        for (var muscle : subMuscles) {
            this.subMuscles.add(muscle);
        }
        for (var muscle : subMuscles) {
            muscle.superMuscles.add(this);
        }
    }

    private Muscle(Muscle sub) {
        this(new Muscle[]{sub});
    }

    private Muscle(String... altNames) {
        this(new Muscle[]{}, altNames);
    }

    private Muscle() {
        this(new Muscle[]{});
    }

    public String getName() {
        return EnumUtils.formatName(super.toString());
    }

    public HashSet<Muscle> getSubMuscles() {
        return this.subMuscles;
    }

    public HashSet<Muscle> getSuperMuscles() {
        return this.superMuscles;
    }

    public boolean matchesName(String name) {
        for (var n : this.altNames) {
            if (n.equalsIgnoreCase(name)) {
                return true;
            }
        }
        return this.getName().equalsIgnoreCase(name);
    }

    public boolean isComposedOf(Muscle m) {
        return Muscle.validateMuscleSuper(this, m);
    }

    public boolean isComponentOf(Muscle m) {
        return Muscle.validateMuscleSub(this, m);
    }

    public boolean isInvolvedIn(Muscle m) {
        return this.isComponentOf(m) || this.isComposedOf(m);
    }

    @Override
    public String toString() {
        return this.getName();
    }

    private static boolean validateMuscleSub(Muscle constant, Muscle dyn) {
        if (constant.matchesName(dyn.getName())) {
            return true;
        }
        for (var subMuscle : dyn.getSubMuscles()) {
            if (Muscle.validateMuscleSub(constant, subMuscle)) {
                return true;
            }
        }
        return false;
    }

    private static boolean validateMuscleSuper(Muscle constant, Muscle dyn) {
        if (constant.matchesName(dyn.getName())) {
            return true;
        }
        for (var superMuscle : dyn.getSuperMuscles()) {
            if (Muscle.validateMuscleSuper(constant, superMuscle)) {
                return true;
            }
        }
        return false;
    }

}

File: exvi-core/src/commonMain/kotlin/com/camackenzie/exvi/core/model/Muscle.kt
/*
 * Copyright (c) Callum Mackenzie 2022.
 */

package com.camackenzie.exvi.core.model

import com.camackenzie.exvi.core.util.EnumUtils
import com.camackenzie.exvi.core.util.SelfSerializable
import kotlinx.serialization.json.*
import kotlinx.serialization.*

/**
 *
 * @author callum
 */
@Serializable
@Suppress("unused")
enum class Muscle(
    subMuscles: Array<Muscle>,
    vararg altNames: String
) : SelfSerializable {
    PectoralisMajor("pecs"),
    Trapezius("traps"),
    Rhomboids(),
    LatissimusDorsi("lats", "latissimus"),
    ErectorSpinae("erector"),
    BicepsBrachii("brachii"),
    BicepsBrachialis("brachialis"),
    Triceps("tricep"),
    Forearms("forearm"),
    AnteriorHead(),
    LateralHead(),
    PosteriorHead(),
    Quadriceps("quads", "quad"),
    Abductors(),
    Adductors(),
    Glutes(),
    ITBand(),
    HipFlexors(),
    Gastrocnemius(),
    Soleus(),
    RecusAbdominis(),
    TranseverseAbdominis(),
    InternalObliques(),
    ExternalObliques(),
    Hamstrings(Abductors, Adductors, ITBand),
    Biceps(BicepsBrachialis, BicepsBrachii),
    LowerBack(ErectorSpinae),
    UpperBack(Trapezius, Rhomboids),
    Neck(),
    Obliques(InternalObliques, ExternalObliques),
    PalmarFascia("grip"),
    PlantarFascia("feet"),
    Back(LowerBack, UpperBack, LatissimusDorsi),
    Chest(PectoralisMajor),
    Arms(Biceps, PalmarFascia, Triceps, Forearms),
    Shoulders(AnteriorHead, LateralHead, PosteriorHead),
    Legs(Quadriceps, Glutes, Hamstrings),
    Calves(Gastrocnemius, Soleus),
    Abs(RecusAbdominis, Obliques, TranseverseAbdominis);

    private val altNames: Array<String>
    val subMuscles = HashSet<Muscle>()
    val superMuscles = HashSet<Muscle>()

    init {
        this.altNames = arrayOf(*altNames)
        for (muscle in subMuscles) {
            this.subMuscles.add(muscle)
        }
        for (muscle in subMuscles) {
            muscle.superMuscles.add(this)
        }
    }

    constructor(vararg altNames: String) : this(emptyArray(), *altNames) {}
    constructor(vararg muscles: Muscle) : this(arrayOf(*muscles))
    constructor() : this(arrayOf<Muscle>()) {}

    val muscleName: String
        get() = EnumUtils.formatName(super.toString())

    fun matchesName(name: String?): Boolean {
        for (n in altNames) {
            if (n.equals(name, ignoreCase = true)) {
                return true
            }
        }
        return this.muscleName.equals(name, ignoreCase = true)
    }

    fun isComposedOf(m: Muscle): Boolean = validateMuscleSuper(this, m)

    fun isComponentOf(m: Muscle): Boolean = validateMuscleSub(this, m)

    fun isInvolvedIn(m: Muscle): Boolean = isComponentOf(m) || isComposedOf(m)

    fun workData(workCoefficient: Double): MuscleWorkData =
        MuscleWorkData(this, workCoefficient)

    override fun toString(): String = muscleName

    override fun getUID(): String = uid

    override fun toJson(): String = Json.encodeToString(this)

    companion object {
        const val uid = "Muscle"

        private fun validateMuscleSub(constant: Muscle, dyn: Muscle): Boolean {
            if (constant.matchesName(dyn.muscleName)) {
                return true
            }
            for (subMuscle in dyn.subMuscles) {
                if (validateMuscleSub(constant, subMuscle)) {
                    return true
                }
            }
            return false
        }

        private fun validateMuscleSuper(constant: Muscle, dyn: Muscle): Boolean {
            if (constant.matchesName(dyn.muscleName)) {
                return true
            }
            for (superMuscle in dyn.superMuscles) {
                if (validateMuscleSuper(constant, superMuscle)) {
                    return true
                }
            }
            return false
        }
    }
}